_blog/2022-12-23-procedural-macros.mdx

```diff
@@ -0,0 +1,417 @@
+        let length = u64::deserialize(iter) as usize;
+        String::from_utf8(iter.take(length).collect()).unwrap()
+    }
+}
+```
+
+> Again this is how to article write not low level byte manipulation. But I know which language I prefer for doing this logic 😄
+### The problem
+Now we have this, we want to add this to a definition for objects
+```rust
+struct Country {
+	name: String,
+	population: u64,
+	above_equator: bool
+}
+```
+
+We could implement `BinarySerializable` for `Country` writing manually in the following way:
+```rust
+impl BinarySerializable for Country {
+    fn serialize(self, buf: &mut Vec<u8>) {
+        self.name.serialize(buf);
+        self.population.serialize(buf);
+        self.above_equator.serialize(buf);
+    }
+
+    fn deserialize<I: Iterator<Item = u8>>(iter: &mut I) -> Self {
+        Self {
+            name: BinarySerializable::deserialize(iter),
+            population: BinarySerializable::deserialize(iter),
+            above_equator: BinarySerializable::deserialize(iter),
+        }
+    }
+}
+```
+
+This is great and we have our desired functionality. In this code we have to be careful we serialize them and deserialize them in the same order. **Writing this `impl` block out for many structs with many fields would get tedious. If we add another `struct` we want to be serializable we don't want to have to have to copy the implementation over. If we change the definition of `Country` we don't want to have to amend this code.**
+### Procedural macro time!
+With some idea of the code we want to generate, we can now make Rust do it for us using a proc macro!
+
+Rust procedural macros require there own special crate for their definition. This is because some of the constraints about how they are compiled. The unofficial Rust convention for derive macros is the name of trait or crate name + derive. So we will run the following from our current folder `cargo new --lib binary-serialize-derive`. With this crate we have to set it be a proc macro by setting the key in its `Cargo.toml`, we will also add dependencies [syn](https://crates.io/crates/syn) for parsing the contents of our structure and [quote](https://crates.io/crates/quote) for generating the output!
+```toml
+[package]
+name = "binary-serialize-derive"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+# Important \/\/\/
+proc-macro = true
+
+[dependencies]
+# Dependencies we use when writing the macro \/\/\/
+quote = "1.0.23"
+syn = "1.0.107"
+```
+
+In our macro we want to parse the input into a structure that we can read information from. In the below we can read `.fields` directly without having to understand what tokens refer to field names and such.
+```rust
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, DeriveInput, Data};
+
+#[proc_macro_derive(BinarySerializable)]
+pub fn my_macro(input: TokenStream) -> TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let name = input.ident;

```

Comment:
I think we should heavily comment the code in this block since it is a tutorial. Eg. here we can say "get the name of the struct"

Reply:
Yes, I don't think it needs too much as it is a little self explanatory. But `ident` probably isn't a commonly used term so it deserves a comment.